internal/pkg/cli/job_deploy.go

```diff
@@ -130,63 +137,104 @@ func (o *deployJobOpts) Execute() error {
 		return err
 	}
 	o.targetEnvironment = env
-
 	app, err := o.store.GetApplication(o.appName)
 	if err != nil {
 		return err
 	}
 	o.targetApp = app
-
 	job, err := o.store.GetJob(o.appName, o.name)
 	if err != nil {
 		return fmt.Errorf("get job configuration: %w", err)
 	}
 	o.targetJob = job
-
 	if err := o.configureClients(); err != nil {
 		return err
 	}
-
 	if err := o.envUpgradeCmd.Execute(); err != nil {
 		return fmt.Errorf(`execute "env upgrade --app %s --name %s": %v`, o.appName, o.targetEnvironment.Name, err)
 	}
-
 	if err := o.configureContainerImage(); err != nil {
 		return err
 	}
+	if err := o.pushToS3Bucket(); err != nil {
+		return err
+	}
+	return o.deployJob()
+}
 
-	addonsURL, err := o.pushAddonsTemplateToS3Bucket()
+func (o *deployJobOpts) pushToS3Bucket() error {
+	mft, err := o.manifest()
 	if err != nil {
 		return err
 	}
+	if err := o.pushEnvFilesToS3Bucket(envFile(o.name, mft)); err != nil {
+		return err
+	}
+	return o.pushAddonsTemplateToS3Bucket()
+}
 
-	return o.deployJob(addonsURL)
+func (o *deployJobOpts) pushEnvFilesToS3Bucket(fileName string) error {
+	if fileName == "" {
+		return nil
+	}
+	content, err := o.ws.ReadSvcFile(o.name, fileName)
+	if err != nil {
+		return fmt.Errorf("read env file %s: %w", fileName, err)
+	}
+	if err := o.retrieveAppResourcesForEnvRegion(); err != nil {
+		return err
+	}
+	reader := bytes.NewReader(content)
+	url, err := o.s3.PutArtifact(o.appEnvResources.S3Bucket, fileName, reader)
```

Comment:
I'm trying to think if replacing an existing env file in the S3 bucket could cause unexpected issues for the customer 💭 (I think we were safe with addons because changeset automatically detect changes)

Like for example: 
1. User updates the .env file in the repository but made no code changes
2. Since there is no change, a new container image is not pushed
3. The env file is uploaded and overrides the existing file in the S3 bucket
4. ⚠️  Deployment fails because change set is empty
5. In the mean time, a new task spins up due to autoscaling but pulls the wrong version of the env file.


Looking at the CDK, if I follow correctly they compute the hash of the file and use that as the key for the object:
https://github.com/aws/aws-cdk/blob/dd5e12d43231d991971de3121928377569b2dd63/packages/%40aws-cdk/aws-s3-assets/lib/asset.ts#L150-L151

I think we should do something similar: https://stackoverflow.com/a/45123990/1201381, compute the sha256 hash of the contents and upload that so that we don't end up in the situation of pulling the wrong file

Reply:
Ooh that's a good catch!

Reply:
ok done